### ISSUE TITLE
Grafana Session should expire when openid token expires (or ask for a refresh token before expiration)

### ISSUE THREAD

**Issue by roidelapluie** [grafana/grafana#12131]
Currently when you log in with openid connect you stay connected to grafana even when the openid token is not valid anymore

**Comment by marefr** [comment]
@DanCech could you identify if this is an actual bug and label the issue correcly? Thanks

**Comment by roidelapluie** [comment]
to me it is a security issue

**Comment by DanCech** [comment]
The current integration doesn't deal with refresh tokens at all, and uses OAuth purely as a login provider.

It would be great to add support for storing the id and refresh token in the user's session and add middleware to deal with refreshing the id token as needed.

There is a fine line between feature and bug here, and you can argue the toss both ways.  There is no doubt that adding that capability would make Grafana more secure against users who have their OAuth account disabled/deleted while logged into Grafana.

I'll take a look into what would be involved to implement this.

**Comment by roidelapluie** [comment]
And checking claims :)

**Comment by seanlaff** [comment]
Hey @DanCech did you have an opportunity to investigate the level of effort required for this? I'm interested in a similar issue that may go hand-in-hand with this one, and willing to lend a hand. https://github.com/grafana/grafana/issues/12556

**Comment by yesoreyeram** [comment]
Does this coming in for 5.5 ?

**Comment by marefr** [comment]
@roidelapluie as of v6.0 you can enforce users to re-authenticate after certain amount of days. Just wanted to mention this since it can help you if you know that your open id provider always expires after certain amount of time. However, I totally understand that it's valuable to have this supported in an automated way. Some more information [here](http://docs.grafana.org/installation/upgrading/#authentication-and-security) and [here](http://docs.grafana.org/auth/overview/#login-and-short-lived-tokens)

**Comment by marefr** [comment]
Should also add that there's an related issue (#15361) open regarding defaults of/possible login duration

**Comment by roidelapluie** [comment]
Thanks @marefr . We are no longer using this feature, we are now using apache mod_auth_openidc to deal with the tokens :)

**Comment by rlcomte** [comment]
Just to be sure: can we expect support for refresh tokens in 6.3 (in combination with OAuth)?
We get questions from our customers who use Grafana on wallboards, which are not too easily accessible.

**Comment by bergquist** [comment]
the 6.3 release have not been fully planned yet so I'm removing it from the milestone.

**Comment by gh7531** [comment]
Is there any status update or even ETA for this feature/issue? 
We are using Grafana with Keycloak for a project and are facing that exact issue.

**Comment by martonorova** [comment]
Hi,
Is there any progress with the refresh token handling issue?
I am also interested in the usecase, when the 'Forward OAuth Identity' is ticked at the data source configuration. It seems that, when the access token is not valid anymore, Grafana does not renew the token with the refresh and stops sending access token to the data source.

I looked it up in the [source code ](https://github.com/grafana/grafana/blob/f2254081c2859e141110c0a948195d9dfb5e5a92/pkg/api/pluginproxy/ds_proxy.go#L321), and in the comment it states, that it handles refreshing the token, but I can't see any proof of that. Am I missing something?

Thanks

**Comment by seanlaff** [comment]
@martonorova From what I understand the `TokenSource` call from the oauth2 lib referenced in this section https://github.com/grafana/grafana/blob/f2254081c2859e141110c0a948195d9dfb5e5a92/pkg/api/pluginproxy/ds_proxy.go#L335-L341 *should* handle refreshing the token if it is expired. Have you observed any error logs from that line below it?

**Comment by martonorova** [comment]
@seanlaff Thanks for the quick answer. After some debug, it turns out that refreshing the access token works fine, my mistake.

The above mentioned problem (Grafana not forwarding the access token to the data source) occurs, if the refresh token expires. In this case, I just get an error like this (using Keycloak):
`Failed to retrieve access token from oauth provider logger=data-proxy-log provider=oauth_generic_oauth error="oauth2: cannot fetch token: 400 Bad Request\nResponse: {\"error\":\"invalid_grant\",\"error_description\":\"Token is not active\"}"`

Shouldn't it take care of refreshing the refresh token as well? Is there a way in Grafana to handle the renewal of the refresh token? Maybe writing a small service for it?

Thanks

**Comment by seanlaff** [comment]
@martonorova Ah, this might be the offending line

https://github.com/grafana/grafana/blob/f2254081c2859e141110c0a948195d9dfb5e5a92/pkg/api/pluginproxy/ds_proxy.go#L347-L354

The entries in the db for `refreshToken`, `accessToken`, `exp`, etc only get updated if the `accessToken` received from TokenSource is different from the one on disk. However if the access token is the same (but a newer refresh token is received), it seems like we may not be writing that new refresh token down.

**Comment by rvansa** [comment]
Have you folks found any workarounds? Since I am embedding the panel in app I consider adding a small iframe invoking `/logout` - with oauth auto login it will immediately log me in again. I think I can't invoke this just from javascript due to CORS and Grafana's missing option to set allowed domains.
Crosslink: https://community.grafana.com/t/embedding-panel-that-forwards-oauth-credentials/34664

**Comment by seanlaff** [comment]
@rvansa and @martonorova How long are the expirations on your access tokens vs refresh tokens?

What I mentioned above is partially to blame for some of this behavior- however fixing it may not actually solve your specific cases. The access token (and refresh token _if_ my bug was to be fixed) are updated on login *and* any request that travels through a datasource with `OAuth Pass Thru`. That leaves a gap though- if a user were to log into grafana via oauth, and stay active *longer* than the expiry of the tokens *without* triggering any datasource requests, we'd end up with a stale token since grafana does not have any daemon-like thing keeping tokens refreshed.  

Generally speaking I think this is somewhat ok as access tokens are shortlived while refresh tokens typically are long (1day+)... making it unlikely to get into this state- but that could be a gross generalization.

@rvansa in regards to your iframe issue, if you host grafana at a subpath `yourapp/grafana`, you can get around that CORS issue.

**Comment by rvansa** [comment]
@seanlaff My access token lifespan is 5 minutes, refresh tokens shouldn't revoked at all (I am using Keycloak). So I am not sure if I am hitting the exact issue; it looks more like the access token not being refreshed before the datasource query (which should be happening according to what you say). Is there any specific logging I should trigger?

Yes, I could spin up a proxy to merge app and Grafana on one port, but I prefer simple setup first.

**Comment by seanlaff** [comment]
@rvansa To be clear, are you leveraging this oauth-pass-thru setting in your datasource? 

https://user-images.githubusercontent.com/3194832/52152295-ef3e1b80-2643-11e9-9025-582d6479034a.png

Unfortunately I don't think there's much logging in the happy-path for this code flow, but if you checkout the errors logs in this function https://github.com/grafana/grafana/blob/f2254081c2859e141110c0a948195d9dfb5e5a92/pkg/api/pluginproxy/ds_proxy.go#L321
you might catch one of those if there is indeed a problem. Can you see if you catch any of those in your output?

Actually- if you turn on debug logging- there should be a log statement for the initial user-login that writes down the token
https://github.com/grafana/grafana/blob/35c67606ce554b7de25c27eebb84140c3f2a9ae4/pkg/services/login/login.go#L180

For reference, this is the PR that originally added the Pass-Thru behavior https://github.com/grafana/grafana/pull/15205

**Comment by rvansa** [comment]
@seanlaff Yes, I am using this option in the datasource.

According to log it really seems the refresh token is expiring:
```
t=2020-08-07T06:42:18+0000 lvl=eror msg="Failed to retrieve access token from oauth provider" logger=data-proxy-log provider=oauth_generic_oauth error="oauth2: cannot fetch token: 400 Bad Request\nResponse: {\"error\":\"invalid_grant\",\"error_description\":\"Refresh token expired\"}"
```

I misunderstood keycloak settings, according to [this SO post](https://stackoverflow.com/questions/52040265/how-to-specify-refresh-tokens-lifespan-in-keycloak) the refresh token lifespan is controller by 'SSO Session Idle' set to 30 minutes by default.

Btw. the iframe approach didn't work well - I've been experiencing errors when the panel was loaded and `/logout` was called at about the same time.

**Comment by seanlaff** [comment]
@rvansa  @martonorova I've opened a PR here https://github.com/grafana/grafana/pull/26885 that I believe will solve your issues with the refresh token

**Comment by rvansa** [comment]
Thanks!

**Comment by weisdd** [comment]
I'm just a beginner in Go and not very familiar with Grafana's architecture, though I think I have figured out how to potentially improve integration with oauth, especially for cases where Grafana stops adding Authorization header to requests to data sources (which is a real issue in environments with short SSO Session Idle timeouts).

[RFC 6759](https://tools.ietf.org/html/rfc6749) defines an exact set of possible Oauth2 errors and the response structure.
Basically, we have [three fields](https://tools.ietf.org/html/rfc6749#section-5.2) (it's `application/json;charset=UTF-8`):

- `error`, required, an error code;
- `error_description`, optional, human-readable text with additional information;
- `error_uri`, optional, A URI identifying a human-readable web page with information about the error.

Amongst the error codes, the most interesting is `invalid_grant`, which means: `The provided authorization grant (e.g., authorization code, resource owner credentials) or refresh token is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or was issued to another client.`.

So, for us, it means that the token, which had been saved in Grafana database, is no longer considered valid. For example, it could happen when `SSO Session Idle` timeout is reached, and, as a result, all tokens are invalidated, including the refresh one. Btw, in Keycloak, the timeout is set to just 30 mins by default. - If a data source validates access token, then a respective user might run into authorization issues a few times a day, after those 30 mins of inactivity.

IMO, we can more or less safely rely on this error code to trigger Grafana session invalidation.

Currently, the token is requested in Proxy's director:

```go
func (proxy *DataSourceProxy) director(req *http.Request) {
...
    if oauthtoken.IsOAuthPassThruEnabled(proxy.ds) {
        if token := oauthtoken.GetCurrentOAuthToken(proxy.ctx.Req.Context(), proxy.ctx.SignedInUser); token != nil {
            req.Header.Set("Authorization", fmt.Sprintf("%s %s", token.Type(), token.AccessToken))
        }
    }
```

That's where we get into trouble. - If `oauthtoken.GetCurrentOAuthToken` returns an error, `director` just does nothing. And it's understandable, it should only rewrite the request.

Let's look inside `GetCurrentOAuthToken` for a moment:

```go
func GetCurrentOAuthToken(ctx context.Context, user *models.SignedInUser) *oauth2.Token {
...
    token, err := connect.TokenSource(ctx, persistedToken).Token()
    if err != nil {
        logger.Error("failed to retrieve OAuth access token", "provider", authInfoQuery.Result.AuthModule, "userId", user.UserId, "username", user.Login, "error", err)
        return nil
    }
```

That's the place where we can catch `invalid_grant`. Below, you could find an example of such an error in action:

```
grafana-deployment-9b599b5fd-q25nj grafana t=2021-02-05T18:23:32+0000 lvl=eror msg="failed to retrieve OAuth access token" logger=oauthtoken provider=oauth_generic_oauth userId=2 username=[REDACTED] error="oauth2: cannot fetch token: 400 Bad Request\nResponse: {\"error\":\"invalid_grant\",\"error_description\":\"Token is not active\"}"
```

Currently, `GetCurrentOAuthToken` does not return any errors, we should change that.

Alright, we got our error, so now we should decide how we can pass it further down. I see two options atm:

- wrap it inside the request context in `director`;
- move the logic for token and Authorization header from `director` to `func (proxy *DataSourceProxy) HandleRequest()`.

I like the latter option as we already have some validation logic there before handing the request over to the proxy backend.

Once we obtain the error as described above, we should parse it as json, and then check the `error_code`. If it's `invalid grant`, then we should invoke something similar to what `logoutUserFromAllDevicesInternal` does:

```go
func (hs *HTTPServer) logoutUserFromAllDevicesInternal(ctx context.Context, userID int64) response.Response {
    userQuery := models.GetUserByIdQuery{Id: userID}

    if err := bus.Dispatch(&userQuery); err != nil {
        if errors.Is(err, models.ErrUserNotFound) {
            return response.Error(404, "User not found", err)
        }
        return response.Error(500, "Could not read user from database", err)
    }

    err := hs.AuthTokenService.RevokeAllUserTokens(ctx, userID)
    if err != nil {
        return response.Error(500, "Failed to logout user", err)
    }

    return response.JSON(200, util.DynMap{
        "message": "User logged out",
    })
}
```

As we can see, we have `AuthTokenService`, which exists within `HTTPServer`, and thus not reachable from `DataSourceProxy` struct:

```go
func NewDataSourceProxy(ds *models.DataSource, plugin *plugins.DataSourcePlugin, ctx *models.ReqContext,
    proxyPath string, cfg *setting.Cfg) (*DataSourceProxy, error) {
    targetURL, err := datasource.ValidateURL(ds.Type, ds.Url)
    if err != nil {
        return nil, err
    }

    return &DataSourceProxy{
        ds:        ds,
        plugin:    plugin,
        ctx:       ctx,
        proxyPath: proxyPath,
        targetUrl: targetURL,
        cfg:       cfg,
    }, nil
}
```

So, now we have two ways:

- either embed proxy into `HTTPServer`;
- expand `DataSourceProxy` struct with `AuthTokenService`.

The first option is probably a bad choice, as it goes against the plugin architecture.

The second option, IMO, is fine. We already pass `*setting.Cfg` from `HTTPServer`, so adding a pointer to `AuthTokenService` should not be a dramatic change.
So, since we see the token is invalid, we can log out the user from all devices or just from this one.

I hope that helps. :)

**Comment by faxm0dem** [comment]
Thanks for this thorough analysis. I'm all for improving the oauth integration. Let's hope this gets addressed with your help.
On a sidenote, I checked https://grafana.com/docs/grafana/latest/administration/configuration/ today and the `duration` options seem to be online in the docs

**Comment by weisdd** [comment]
@faxm0dem Yesterday, one of Grafana's team members merged [the PR](https://github.com/grafana/grafana/pull/31000) for documentation improvement regarding `duration` within an hour after it was opened, that's why you saw it there today :)

**Comment by faxm0dem** [comment]
This oauth stuff is really very confusing. I have many problems with datasources (having `Forward OAuth Identity` checked) but the keycloak admins say everything's fine on their side. So I'm guessing grafana must do something wrong here. So I'm playing with all those `lifetime` and `token` and `cookie` settings like the dog who doesn't know what it's doing :rofl:

**Comment by roidelapluie** [comment]
This is not a magic thing ; it is very possible that backends are looking for the audience of the token and might reject it even if valid.

**Comment by faxm0dem** [comment]
I'm not expecting it to be magic (although that would be very cool), but I'm realising that I'm a bit off topic. This issue you opened is not directly related to datasources, just with being still logged in while the openid token is invalid.
That being said, it also has the side effect that it's the datasources that get forwarded that token that send error messages to the user. So the user is still connected, but the datasource won't send any data so all the panels will complain.

**Comment by weisdd** [comment]
@faxm0dem I totally agree with what you're saying :)

We're currently building an internal PaaS with SSO based on Keycloak with aim to have a role-based authorization everywhere.
And I was surprised to see how difficult it is to implement authorization for metrics.

Had to develop something similar to [prometheus-acls](https://github.com/bitsbeats/prometheus-acls) for rewriting requests to Prometheus / VictoriaMetrics. I chose [VictoriaMetrics/metricsql](https://github.com/VictoriaMetrics/metricsql) to play with label filters as this library is a simple and lightweight alternative to the Prometheus library.
And then all troubles with jwt tokens began:
1. [grafana-operator](https://github.com/integr8ly/grafana-operator) did not have support for setting `oauthPassThru: true` for datasources.
    Had to implement it ([PR](https://github.com/integr8ly/grafana-operator/pull/343)).
2. Had no idea how to validate jwt in Go.
    Spent some time reading about it.
3. Keycloak did not include my `client` in `aud`, so [coreos/go-oidc](https://github.com/coreos/go-oidc) failed to validate access tokens.
    Had to add a mapper.
3. Found out that grafana would stop forwarding the token after a while.
    After reading through RFCs, grafana's source code, and its documentation, just increased `SSO Session Idle` and `SSO Session Max` in Keycloak to `25h` and decreased `login_maximum_inactive_lifetime_days` and `login_maximum_lifetime_days` in Grafana to `1` day. Everything got better, though not fully. - I guess the gap of 1h was not enough. 
    Fortunately, I found the `login_maximum_lifetime_duration` and `login_maximum_inactive_lifetime_duration` options, and wanted to set them to `8h` and `4h` respectively... But grafana-operator did not have support for that, so another [PR](https://github.com/integr8ly/grafana-operator/pull/345).
4. For proxying web-interface for Prometheus, I chose [oauth2-proxy](https://github.com/oauth2-proxy/oauth2-proxy). Should have been simple, right?
    - oauth2-proxy puts id token instead of access token to the Authorization header (`OAUTH2_PROXY_PASS_AUTHORIZATION_HEADER=true`), thus can't rely on that;
    - fortunately, `OAUTH2_PROXY_PASS_ACCESS_TOKEN=true` makes it add the token to `X-Forwarded-Access-Token` or `X-Auth-Request-Access-Token` - had to add the respective logic to my tool;
    - oauth2-proxy can send stale access tokens (in Keycloak, they're short-lived by default), so we need to instruct the proxy for updating cookies more often (`OAUTH2_PROXY_COOKIE_REFRESH=3m`), but guess what - `keycloak` provider does not support the option, so we need to use `oidc` instead.

Those are just things that were on top of my head atm.

To summarize, yep, I feel the same and also wish it was way simpler than described above and worked well with just one magic button :)
As for now, I hope there's a community member amongst us who's good at Go and is willing to make a PoC for oauth2 improvement in Grafana. I wish I could do it myself, though I started looking into Go only about a month ago (came from scripting in Python), thus certain things in Go still look like rocket science to me :)

**Comment by lovasoa** [comment]
Hello ! 
  We are still seeing requests sent by data source proxies without an authorization header after the oauth "SSO Session Idle" timeout, even with the latest grafana token refresh enabled.

   Looking at the code, the issue [described by @weidd above](https://github.com/grafana/grafana/issues/12131#issuecomment-775112877) still seems to be present in the latest code (ignoring errors and silently dropping the authorization header [here](https://github.com/grafana/grafana/blob/v10.1.1/pkg/api/pluginproxy/ds_proxy.go#L255C1-L255C114)).

Could we re-open this issue ?

**Comment by mgyongyosi** [comment]
Hey @lovasoa, have you tried with the latest version on `main`? We have made improvements to the refresh token handling/expiration check which will be released with Grafana 10.2.  (https://github.com/grafana/grafana/pull/69829#issuecomment-1647821813)

If this issue is still a thing with the latest version, could you please open a separate issue, because this one is mainly about token expiry and your issue is related to the `Forward OAuth Identity`? 

Thanks!
